#BSUB -P RNASeq_pipelines
#BSUB -n 1
#BSUB -M 2000
#BSUB -oo 00_preprocess.out -eo 00_preprocess.err
#BSUB -J 00_preprocess
#BSUB -q standard

## 1. Preprocess to prepare the RAW FASTQ
## 1.1 If you have multiple fastq files for each sample, usually generated by different lanes, you need to merge them into one.

# For single-end sequencing
cat sample1_L001.fq.gz sample1_L002.fq.gz sample1_L003.fq.gz sample1_L004.fq.gz > sample1.raw.fq.gz

# For paired-end sequencing
cat sample1_L001_R1.fq.gz sample1_L002_R1.fq.gz sample1_L003_R1.fq.gz sample1_L004_R1.fq.gz > sample1_R1.raw.fq.gz
cat sample1_L001_R2.fq.gz sample1_L002_R2.fq.gz sample1_L003_R2.fq.gz sample1_L004_R2.fq.gz > sample1_R2.raw.fq.gz

## 1.2 If you start from BAM files, usually downloaded from some databases, you need to convert them into FASTQ files.

bedtools=/research/rgs01/applications/hpcf/apps/bedtools/install/2.25.0/bin/bedtools
$bedtools bamtofastq -i input.bam -fq output.fqi # For single-end sequencing
$bedtools bamtofastq -i input.bam -fq output_R1.fq -fq2 output_R2.fq # # For paired-end sequencing

