# RNASeq_Quantification_Pipelines

## Overview
![Picture1](https://user-images.githubusercontent.com/33663247/66729168-3ef9f700-ee0f-11e9-866a-d4a621466c9d.png)


### 0. Preprocess to prepare the RAW FASTQ

* If you have multiple fastq files for each sample, usually generated by different lanes, you need to merge them into one.
* If you start from BAM files, usually downloaded from some databases, you need to convert them into FASTQ files.


### 1. Quality Control of RAQ FASTQ Files by FastQC

Note: From this analysis, we need to figure out:
a. **Encoding of Phred Scores**: Phred+33 is listed as Illumina 1.9/Sanger, while Phred+64 encoding as illumina 1.5 or lower. (Find more details here: https://sequencing.qcfail.com/articles/incorrect-encoding-of-phred-scores/)
b. **Sequence Length**: The most common values are 46/45, 76/75, 101/100 or 151/150.
c. **Adapter Type**: Illumina Universal Adapter(AGATCGGAAGAG), Illumina Small RNA 3' Adapter(TGGAATTCTCGG), Illumina Small RNA 5' Adapter(GATCGTCGGACT), Nextera Transposase Sequence(CTGTCTCTTATA) and SOLID Small RNA Adapter(CGCCTTGGCCGT).
