# RNASeq_Quantification_Pipelines

![Picture1](https://user-images.githubusercontent.com/33663247/66729168-3ef9f700-ee0f-11e9-866a-d4a621466c9d.png)


## 0. Preprocess to prepare the RAW FASTQ

0.1 If you have multiple fastq files for each sample, usually generated by different lanes, you need to merge them into one.

##### For single-end sequencing
  cat sample1_L001.fq.gz sample1_L002.fq.gz sample1_L003.fq.gz sample1_L004.fq.gz > sample1.raw.fq.gz

##### For paired-end sequencing
  cat sample1_L001_R1.fq.gz sample1_L002_R1.fq.gz sample1_L003_R1.fq.gz sample1_L004_R1.fq.gz > sample1_R1.raw.fq.gz
  cat sample1_L001_R2.fq.gz sample1_L002_R2.fq.gz sample1_L003_R2.fq.gz sample1_L004_R2.fq.gz > sample1_R2.raw.fq.gz

### 0.2 If you start from BAM files, usually downloaded from some databases, you need to convert them into FASTQ files.

  bedtools=/research/rgs01/applications/hpcf/apps/bedtools/install/2.25.0/bin/bedtools
  $bedtools bamtofastq -i input.bam -fq output.fqi # For single-end sequencing
  $bedtools bamtofastq -i input.bam -fq output_R1.fq -fq2 output_R2.fq # # For paired-end sequencing

