---
title: Alignment-free quantification by Salmon
layout: default
nav_order: 1
permalink: /docs/quantification/alignment_free/alignment_free
parent: Quantification
---

Salmon is a bulk RNA-Seq quantifier with **wicked-fast speed** and **comparable accuracy** (https://salmon.readthedocs.io/en/latest/salmon.html). It provides two working modes:

* **Mapping-based mode**: this is the feature mode that makes Salmon famous. Samlon employes a SEME (super maximal exact match)-based chaining algorithm to find and score potential mapping loci, making it super fast (because no alignment is needed) and comprably accurate. From version 1.0.0 and higher, Salmon introduced the **Selective Alignment** and **Decoy Sequences** to further improve its quantification accuracy (for more details: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02151-8).
* **Alignment-based mode**: Salmon is not an aligner, so this mode, though called alignment-based, does NOT take FASTQ files. Instead, you can simpley provide: 1) BAM/SAM files of alignments generated by other aligners, e.g., STAR, Bowtie; and 2) FASTA file of reference transcriptome. No indexing is required for this mode.

In this pipeline, we suggest to use the mapping-based mode ONLY.

```bash
## Indexing directories
index_hg38v43=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg38/gencode.release43/bulkRNAseq/Salmon/index_decoy
index_hg19v43=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg19/gencode.release43/bulkRNAseq/Salmon/index_decoy
index_mm39vM32=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm39/gencode.releaseM32/bulkRNAseq/Salmon/index_decoy
index_mm10vM25=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm10/gencode.releaseM25/bulkRNAseq/Salmon/index_decoy

## transcript-to-gene mapping files
tr2gene_hg38v43=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg38/gencode.release43/gencode.v43.primary_assembly.annotation.gtf
tr2gene_hg19v43=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg19/gencode.release43/gencode.v43lift37.annotation.gtf
tr2gene_mm39vM32=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm39/gencode.releaseM32/gencode.vM32.primary_assembly.annotation.gtf
tr2gene_mm10vM25=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm10/gencode.releaseM25/gencode.vM25.primary_assembly.annotation.gtf

## Quantification by Salmon
# For paired-end sequencing
salmon quant -i $hg38v43_index -l A -p 8 -g $tr2gene_hg38v43 -1 /your_path/fqClean_R1.fq.gz -2 /your_path/fqClean_R2.fq.gz --validateMappings -o /your_path/quantSalmon
# For single-end sequencing
salmon quant -i $hg38v43_index -l A -p 8 -g $tr2gene_hg38v43 -r /your_path/fqClean.fq.gz --validateMappings -o /your_path/quantSalmon
```

**<u>Key arguments:</u>**

Salmon is very smart, It could learn most of the arguments by itseft. So there are only a few arguments you need to specify, making it super easy to use.

* **-l/--libType**: library type. Salmon employs a three-letter string to denote the library type (https://salmon.readthedocs.io/en/latest/library_type.html#fraglibtype): 1) the relative orientation of two matched mates, including I = inward, O = outward and M = matching; 2) the protocol is stranded or unstranded, including S = stranded and U = unstranded; 3) the strand from which the read originates in a strand-specific protocol, including F = read 1 (or single-end read) from the forward stand and R =  read 1 (or single-end read) from the reverse stand. As a result, there are 9 library types: **3 for unstranded library (IU, MU, OU)** and **6 for stranded library (ISF, ISR, MSF, MSR, OSR, OSF)**, as shown below. Unless you know the exact library type of your samples, just use "A", which asks Salmon to figure it out automatically.

  ![_images/ReadLibraryIllustration.png](https://salmon.readthedocs.io/en/latest/_images/ReadLibraryIllustration.png)

* -i/--index: salmon index files. We have generated them for hg38, hg19, mm39 and mm10. You don't need to regenerate them.

* -g/--geneMap: file contaning a mapping of transcript to gene. **This must be provided to genereate the gene-level abundance estimates.**

* -p/--threads: number of threads to use concurrently.

<u>**Key outputs**</u>: (for more details: https://salmon.readthedocs.io/en/latest/file_formats.html#fileformats)

* **quant.sf:** contains transcript-level abundance estimates, TPM and raw counts.

* **quant.genes.sf**: contains the aggregated gene-level abundance estimates, TPM and raw counts.

* **lib_format_counts.json**: this file contains a count of the number of mappings that matched each possible library type. You can find the infered library type from the line of "expected_format". Here is a summary of library types used in different tools:

  | Salmon (--libType, SE/PE) | RSEM (--strandedness) | TopHap (--library-type) | HTSeq (--stranded) |
  | ------------------------- | --------------------- | ----------------------- | ------------------ |
  | U/IU                      | none                  | -fr-unstranded          | no                 |
  | SR/ISR                    | reverse               | -fr-firststrand         | reverse            |
  | SF/ISF                    | forward               | -fr-secondstrand        | yes                |

  