#BSUB -P RNASeq_pipelines
#BSUB -n 1
#BSUB -M 2000
#BSUB -oo 00_preprocess.out -eo 00_preprocess.err
#BSUB -J 00_preprocess
#BSUB -q standard

## 1. Preprocess to prepare the RAW FASTQ
## 1.1 If you have multiple fastq files for each sample, usually generated by different lanes, you need to merge them into one.

# For single-end sequencing
cat sample1_L001.fq.gz sample1_L002.fq.gz sample1_L003.fq.gz sample1_L004.fq.gz > sample1.raw.fq.gz

# For paired-end sequencing
cat sample1_L001_R1.fq.gz sample1_L002_R1.fq.gz sample1_L003_R1.fq.gz sample1_L004_R1.fq.gz > sample1_R1.raw.fq.gz
cat sample1_L001_R2.fq.gz sample1_L002_R2.fq.gz sample1_L003_R2.fq.gz sample1_L004_R2.fq.gz > sample1_R2.raw.fq.gz

## 1.2 If you start from BAM files, usually downloaded from some databases, you need to convert them into FASTQ files.

bedtools=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_apps/apps/bedtools2/bin/bedtools

$bedtools bamtofastq -i input.bam -fq output.raw.fq # For single-end sequencing
$bedtools bamtofastq -i input.bam -fq output_R1.raw.fq -fq2 output_R2.raw.fq # # For paired-end sequencing

# NOTE:
# 1) The BAM files of paired-end sequencing, MUST BE SORTED by NAME. To sort the BAM files, please use "samtools sort -n -o output.sorted input.bam". [* IF Error: fail to open file output.sorted, please use "samtools sort -n input.bam output.sorted"]
# 2) Duplication issue. Multiple alignments (>1 hits for a single read) may exist in the BAM files. In this case, bedtools (v2.25 and earlier) would parse each alignment and generate two mates. As a results, the reads with multiple alignments would be duplicated in the fastq files. The duplicateion issue would cause errors in RSEM analysis and improper quantificaiton in Salmon analysis (the expression of transcripts/genes with duplicated reads would be over-estimated). The author has fixed this problem in the version of 2.29 and later.
