#BSUB -P RNASeq_pipelines
#BSUB -n 8
#BSUB -M 3000
#BSUB -oo 03_quantRSEM.out -eo 03_quantRSEM.err
#BSUB -J 03_quantRSEM
#BSUB -q standard

# software
rsem=/research_jude/rgs01_jude/applications/hpcf/apps/RSEM/1.3.0/rsem-calculate-expression
bowtie2=/research_jude/rgs01_jude/applications/hpcf/apps/bowtie/install/2.2.9/bin
bedtools=/research_jude/rgs01_jude/applications/hpcf/apps/bedtools/install/2.25.0/bin/bedtools
genebody=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/git_repo/RNASeq_pipelines/05_genebodyCoverage.R

# database
reference_hg38bowtie2=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg38/gencode.release32/RSEM/index_bowtie2/hg38
reference_hg38star=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg38/gencode.release32/RSEM/index_star/hg38
reference_hg19bowtie2=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg19/gencode.release32/RSEM/index_bowtie2/hg19
reference_hg19star=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg19/gencode.release32/RSEM/index_star/hg19
reference_mm10bowtie2=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm10/gencode.releaseM23/RSEM/index_bowtie2/mm10
reference_mm10star=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm10/gencode.releaseM23/RSEM/index_star/mm10
binlist_hg38=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg38/gencode.release32/binlist_150.txt
binlist_hg19=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/hg19/gencode.release32/binlist_150.txt
binlist_mm10=/research_jude/rgs01_jude/groups/yu3grp/projects/software_JY/yu3grp/yulab_databases/references/mm10/gencode.releaseM23/binlist_150.txt

# I/O
indir=
outdir=

## For single-end sequencing
# Bowtie2 Alignment and RSEM Quantification
$rsem -p 8 --bowtie2 --bowtie2-path $bowtie2 --bowtie2-sensitivity-level sensitive --strandedness reverse --sort-bam-by-coordinate --phred33-quals $indir/sample.clean.fq.gz $reference_hg38bowtie2 $outdir/sample/quant

# Gene Body Coverage
$bedtools multicov -bams $outdir/sample/quant.transcript.sorted.bam -bed $binlist_hg38 > $outdir/sample/readsDistribution.txt
$genebody $outdir/sample/readsDistribution.txt $outdir/sample/genebodyCoverage

## For paired-end sequencing
# Bowtie2 Alignment and RSEM Quantification
$rsem -p 8 --bowtie2 --bowtie2-path $bowtie2 --bowtie2-sensitivity-level sensitive --strandedness reverse --sort-bam-by-coordinate --phred33-quals --paired-end $indir/sample_R1.clean.fq.gz $indir/sample_R2.clean.fq.gz $reference_hg38bowtie2 $outdir/sample/quant

# Gene Body Coverage
$bedtools multicov -bams $outdir/sample/quant.transcript.sorted.bam -bed $binlist_hg38 > $outdir/sample/readsDistribution.txt
$genebody $outdir/sample/readsDistribution.txt $outdir/sample/genebodyCoverage

# Options to set
# 1) --strandedness: none|forward|reverse. You can get this information from the one who prepared the libraries. Otherwise, you need to figure it out yourself. Salmon is recommended. The standard outputs of Salmon, lib_format_counts.json, provides this information.
# 2) sorting method: --sort-bam-by-coordinate or --sort-bam-by-read-name.
# 3) BAM Outputs: BAM file of transcriptome is generated by default. Use --no-bam-output to cancal it, or use --output-genome-bam to generate the BAM file of genome simultaneously.
# 4) Encoding of Phred Scores: --phred33-quals or --phred64-quals. Check the FastQC output: Phred+33 is listed as Illumina 1.9/Sanger, while Phred+64 encoding as illumina 1.5 or lower.
# 5) For more information, please check out here: https://github.com/bli25broad/RSEM_tutorial.

# Key Outputs:
# 1) quant.genes.results: quantification results of transcripts/isoforms. TPM, FPKM and COUNTS are provided.
# 2) quant.isoforms.results: quantification results of genes. TPM, FPKM and COUNTS are provided.
# 3) quant.transcript.bam: alignments to reference transcriptome
# 4) quant.transcript.sorted.bam: sorted alignments to reference transcriptome
# 5) quant.transcript.sorted.bam.bai: index of sorted alignments to reference transcriptome


